File: src/main/java/com/vc/medium/L216.scala
package com.vc.medium

object L216 {
  def combinationSum3(k: Int, n: Int): List[List[Int]] = {
    import scala.collection.mutable
    val res = new mutable.ListBuffer[List[Int]]()
    val list = new mutable.ListBuffer[Int]()
    def solve(start:Int, remain: Int, count: Int): Unit = {
      if(remain == 0 && count == 0) res += list.toList
      else if(remain > 0 && count > 0) {
        (start to 9).foreach(i => {
          list += i
          solve(i + 1, remain - i, count - 1)
          list.remove(list.length - 1)
        })
      }
    }
    solve(1, n, k)
    res.toList
  }
}

File: src/main/java/com/vc/medium/L216.java
package com.vc.medium;

import java.util.*;

class L216 {

    private List<List<Integer>> res;
    private boolean[] visited;
    private int k, n;

    public List<List<Integer>> combinationSum3(int k, int n) {
        this.res = new ArrayList<>();
        this.visited = new boolean[9];
        this.k = k;
        this.n = n;
        helper(0, 0, new ArrayList<>());
        return res;
    }

    private void helper(int index, int current, List<Integer> list) {
        if(list.size() == k) {
            if(current == n) res.add(new ArrayList<>(list));
        }

        for(int i = index; i < visited.length; i++) {
            if(visited[i]) continue;

            list.add(i + 1);
            visited[i] = true;
            helper(i, current + i + 1, list);
            visited[i] = false;
            list.remove(list.size() - 1);
        }
    }
}
